File: scripts/defunctParser.py
import yaml
from datetime import datetime, timedelta

def parse_yaml_file(file_path):
    with open(file_path, 'r') as file:
        data = yaml.safe_load(file)
    data = {str(key): value for key, value in data.items()}
    return data

def get_old_app_ids(yaml_data, days_threshold=15):
    old_app_ids = []
    current_date = datetime.now()
    date_threshold = current_date - timedelta(days=days_threshold)
    for date_str, app_ids in yaml_data.items():
        if app_ids is not None: 
            date_added = datetime.strptime(date_str, '%Y-%m-%d')
            if (current_date - date_added).days >= days_threshold:
                cleaned_app_ids = [app_id.replace('_', '', 1) for app_id in app_ids]
                old_app_ids.extend(cleaned_app_ids)
    return old_app_ids, date_threshold.strftime('%Y-%m-%d')

def format_app_ids(app_ids):
    return ','.join(app_ids)

if __name__ == "__main__":
    yaml_file_path = '_data/defunct.yaml'
    data = parse_yaml_file(yaml_file_path)
    
    old_app_ids, date_threshold = get_old_app_ids(data)
    formatted_app_ids = format_app_ids(old_app_ids)
    print(formatted_app_ids)

File: scripts/defunctParser.js
const fs = require('fs');
const yaml = require('js-yaml');
const path = require('path');

function parseYamlFile(filePath) {
    const fileContents = fs.readFileSync(filePath, 'utf8');
    let data;
    try {
        data = yaml.load(fileContents);
    } catch (e) {
        console.error(`Failed to parse YAML file: ${e.message}. \n\nFind the problematic dates and try remove them first`);
        process.exit(1)
    }
    
    const parsedData = {};
    for (const key in data) {
        if (data.hasOwnProperty(key)) {
            if (parsedData.hasOwnProperty(key)) {
                console.warn(`Duplicate key found: ${key}. Merging entries.`);
                parsedData[key] = [...new Set([...parsedData[key], ...data[key]])];
            } else {
                parsedData[key] = data[key];
            }
        }
    }
    return parsedData;
}

function formatDate(date) {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const day = String(date.getDate()).padStart(2, '0');
    return `${year}-${month}-${day}`;
}

function getOldAppIds(yamlData, daysThreshold = 15) {
    const oldAppIds = [];
    const currentDate = new Date();
    const dateThreshold = new Date(currentDate - daysThreshold * 24 * 60 * 60 * 1000);
    for (const dateStr in yamlData) {
        if (yamlData.hasOwnProperty(dateStr) && yamlData[dateStr] !== null && yamlData[dateStr].length > 0) {
            const dateAdded = new Date(dateStr);
            if ((currentDate - dateAdded) / (1000 * 60 * 60 * 24) >= daysThreshold) {
                const cleanedAppIds = yamlData[dateStr].map(appId => appId.replace('_', ''));
                oldAppIds.push(...cleanedAppIds);
                delete yamlData[dateStr];
            }
        }
    }
    return { oldAppIds, dateThreshold: formatDate(dateThreshold), updatedYamlData: yamlData };
}

function formatAppIds(appIds) {
    return appIds.join(',');
}

function writeYamlFile(filePath, data) {
    const formattedData = {};
    for (const key in data) {
        if (data.hasOwnProperty(key) && data[key] !== null && data[key].length > 0) {
            formattedData[formatDate(new Date(key))] = data[key];
        }
    }

    let yamlStr = yaml.dump(formattedData, { lineWidth: -1, quotingType: '"', forceQuotes: false, indent: 2 });
    yamlStr = yamlStr.replace(/"(\d{4}-\d{2}-\d{2})"/g, '$1');

    yamlStr = yamlStr.replace(/^(\s*)-\s/gm, '- ');

    fs.writeFileSync(filePath, yamlStr, 'utf8');
}

const yamlFilePath = path.join(__dirname, '../_data/defunct.yaml');
const data = parseYamlFile(yamlFilePath);

const { oldAppIds, dateThreshold, updatedYamlData } = getOldAppIds(data);
const formattedAppIds = formatAppIds(oldAppIds);

console.log(formattedAppIds);

writeYamlFile(yamlFilePath, updatedYamlData);
